libsnark/zk_proof_systems/plonk/prover.hpp

```diff
@@ -0,0 +1,252 @@
+        Field &r_zeta)
+        : W_polys_blinded_at_secret_g1(W_polys_blinded_at_secret_g1)
+        , z_poly_at_secret_g1(z_poly_at_secret_g1)
+        , t_poly_at_secret_g1(t_poly_at_secret_g1)
+        , a_zeta(a_zeta)
+        , b_zeta(b_zeta)
+        , c_zeta(c_zeta)
+        , S_0_zeta(S_0_zeta)
+        , S_1_zeta(S_1_zeta)
+        , z_poly_xomega_zeta(z_poly_xomega_zeta)
+        , W_zeta_at_secret(W_zeta_at_secret)
+        , W_zeta_omega_at_secret(W_zeta_omega_at_secret)
+        , r_zeta(r_zeta)
+    {
+    }
+};
+
+// Prover round 0 output
+template<typename ppT> struct round_zero_out_t {
+    // - zh_poly: vanishing polynomial Zh (from round 0)
+    std::vector<libff::Fr<ppT>> zh_poly;
+    // - null_poly: 0 polynomial (from round 0)
+    polynomial<libff::Fr<ppT>> null_poly;
+    // - neg_one_poly: -1 polynomial (from round 0)
+    polynomial<libff::Fr<ppT>> neg_one_poly;
+};
+
+// Prover round 1 output
+template<typename ppT> struct round_one_out_t {
+    // - blind_scalars: blinding scalars b1, b2, ..., b9 (only
+    //   b1-b6 used in round 1)
+    std::vector<libff::Fr<ppT>> blind_scalars;
+    // - W_polys: witness polynomials (Lagrange interpolation of the
+    //   witness values)
+    std::vector<polynomial<libff::Fr<ppT>>> W_polys;
+    // - W_polys_blinded: blinded witness polynomials
+    std::vector<std::vector<libff::Fr<ppT>>> W_polys_blinded;
+    // - W_polys_blinded_at_secret_g1: the blinded witness polynomials
+    //   evaluated at the secret input denoted [a]_1, [b]_1, [c]_1 in
+    //   [GWC19]
+    std::vector<libff::G1<ppT>> W_polys_blinded_at_secret_g1;
+};
+
+// Prover round 2 output
+template<typename ppT> struct round_two_out_t {
+    // - beta: permutation challenge -- hashe of transcript
+    libff::Fr<ppT> beta;
+    // - gamma: permutation challenge -- hashe of transcript
+    libff::Fr<ppT> gamma;
+    // - z_poly: blinded accumulator poly z(x)
+    polynomial<libff::Fr<ppT>> z_poly;
+    // - z_poly_at_secret_g1: blinded accumulator poly z(x) evaluated at
+    // secret
+    libff::G1<ppT> z_poly_at_secret_g1;
+};
+
+// Prover round 3 output
+template<typename ppT> struct round_three_out_t {
+    // - alpha: quotinet challenge -- hash of transcript
+    libff::Fr<ppT> alpha;
+    // - z_poly_xomega: the polynomial z(x*w) i.e. z(x) shifted by w
+    std::vector<libff::Fr<ppT>> z_poly_xomega;
+    // - t_poly: t(x) divided in three parts t(x) = t_lo(x) + t_mid(x) x^n
+    //   + t_hi(x) x^{2n}
+    std::vector<polynomial<libff::Fr<ppT>>> t_poly;
+    // - t_poly_long: the quotient polynomial t(x) (see Round 3, pp28
+    //   [GWC19])
+    polynomial<libff::Fr<ppT>> t_poly_long;
+    // - t_poly_at_secret_g1: t(x) evaluated at the secret input zeta
+    //   i.e. t(zeta)
+    std::vector<libff::G1<ppT>> t_poly_at_secret_g1;
+};
+
+// Prover round 4 output
+template<typename ppT> struct round_four_out_t {
+    // - zeta: evaluation challenge -- hash of transcript
```

Comment:
We may want these structs to have constructors with move semantics for efficient initialization.  This is how things usually work for structures in the existing proof schemes.